### PR TITLE
Fix undefined method gsub error with attachment text attribute is nil

### PR DESCRIPTION
**Pull request by markquezada**
If an attachment's `text` attribute is `nil`, `sub_html_links` in `link_formatter.rb` fails with `undefined method 'gsub' for nil:NilClass`. There are probably many different ways to solve this but it seemed most beneficial to me to simply remove attributes that are set to `nil` before processing. Unfortunately, there is no built-in `compact` method for hashes so one is provided here.

For context: our use case includes sending ourselves slack notifications of some types of user-generated content. Without this change, we'd have to conditionally build the individual attachment payloads to prevent setting `text` keys with `nil` values. With this change, `nil` values are simply ignored by the slack notifier library so much less conditional logic is necessary to build slack payloads.

Happy to answer questions or do additional updates if necessary.

### CODE REVIEW

**Comment by markquezada** [comment]
I'm actually not sure this is working as intended as I'm still seeing this error in some cases. Probably need to expand on the test/failure cases.

**Comment by stevenosloan** [comment]
**tl;dr;** thanks for the report, that's definitely an issue -- closing in favor of https://github.com/stevenosloan/slack-notifier/pull/82 that should fix the issue in a more general sense.

Awesome, thanks @markquezada  -- sorry to take so long to get back to this, I've looked a few times and never had enough time to properly respond 😬. This is definitely an issue, thanks for bringing it up -- and your test cases were directly useful 😄, but for a fix I think I'm prefering checking for the "string"'s ability to respond to `gsub` in the link formatter. This willl help in the case a middleware has introduced (or is expecting) a nil or other object that doesn't support gsub.

As for the inconsistency you've seen, I think that's coming from the fact in your `recursive_compact` you're only checking for hashes, and not handling arrays. Attachments can (and should really) be passed as an array of hashes so it could still fail for you in that case. Looking into that though, I found another bug so cheers for pointing it out!

I'm closing this in favor of https://github.com/stevenosloan/slack-notifier/pull/82 which uses the "gsub check" fix, if you have a chance to check that out it'd be great.

**Comment by markquezada** [comment]
Good catch on the array/hashes issue. 👍 

#82 Looks like a great solution. Thanks for fixing this!

**Comment by stevenosloan** [comment]
Cool, got that merged to master -- will cut a release once I'm able to get some eyes on the other PR that came out of these changes.
